### PR TITLE
Changes in driver adhering to New JSON API Server Spec

### PR DESCRIPTION
**Pull request by kathirsvn**
The JSON API server (i.e. new version of the docs server) has new http specifications as outlined here.

https://github.com/stargate/doc-api/blob/main/document-api-spec.textile
https://github.com/stargate/doc-api/blob/main/document-api-network-spec.textile

Stargate mongoose driver (forked from [stargate/stargate-mongoose](https://github.com/stargate/stargate-mongoose)) repo is being modified in this repo to adhere to the new specification.

The operations (such as insertOne, findOne, find, etc.) that are agreed for development are tracked in a 'Postman' collection here. 
https://datastax.postman.co/workspace/Stargate-Doc-API-V3-Dev~641686bd-5f9d-4def-b0f8-47fd562868ed/collection/24363330-6ec27e5b-b831-4997-89ca-34e361039d40?action=share&creator=24381626

Below are the APIs modified as per the new specification.

The old implementations of the APIs that are yet to be implemented (for example `findOneAndReplace`) have been removed, to have a clean implementation once those are agreed for development as well. 

_Note : 'deleteMany' API is yet to be implemented on the server side, and so, instead of failing, it has been set to return empty result (i.e. success) for now._

In order to run this branch locally, we need to set up `.env` file based on the `.env.example` file and I can share the postman details separately if you would like to run the tests against postman.

| API             |  
| ----------- | 
| find
| findOne
| insertOne
| insert (internally invokes insertMany)
| insertMany
| findOneAndUpdate
| deleteOne
| updateOne
| updateMany
| update (internally invokes updateMany)

### CODE REVIEW

**Comment by kidrecursive** [comment]
Heya, this looks good, is the plan to keep it in this repo until the document API is further along in stargate?

**Review by kidrecursive — APPROVED** [review]

File: src/client/httpClient.ts
Comment (amorton) [inline src/client/httpClient.ts]: NIT: remove old code in all places.

File: src/client/httpClient.ts
Comment (amorton) [inline src/client/httpClient.ts]: The error message is swallowed, lets set message to be `e.message` and also add `error:e` to the object?

File: src/client/httpClient.ts
Comment (amorton) [inline src/client/httpClient.ts]: logging should be using the `logger` imported at the top of the file and then log at the appropriate level, e.g. 

```
logger.debug(`Request url ${url}`);
```
If possible we want to avoid the string building if the debug log level is not active...

```
logger.debug("_request with URL %s", requestInfo.url);
```
Check with @kidrecursive if stuck , see https://stackoverflow.com/questions/56561576/string-interpolation-with-winston-not-working-in-nodejs

File: src/collections/cursor.ts
Comment (amorton) [inline src/collections/cursor.ts]: remove reference to mongo pls :)

File: src/collections/cursor.ts
Comment (amorton) [inline src/collections/cursor.ts:29]: nit: confusing to have projection, it's part of the query. But I now understand the query property here is only the `filter` of the query. Consider renaming `query` to be `filter` to make clearer.

File: src/collections/cursor.ts
Comment (amorton) [inline src/collections/cursor.ts]: nit: collection data type should be Collection not any

File: src/collections/cursor.ts
Comment (amorton) [inline src/collections/cursor.ts]: nit: change batchSize to pageSize if they are the same thing. It should match the name used in command options.

File: src/collections/cursor.ts
Comment (amorton) [inline src/collections/cursor.ts]: nit: use the same name for the pageState

File: src/collections/db.ts
Comment (amorton) [inline src/collections/db.ts]: Do we need to catch if we are not doing anything with the error.

File: src/collections/utils.ts
Comment (amorton) [inline src/collections/utils.ts]: check if this function is still needed, it may be a hangover from the initial R&D version we made - from talking with @kidrecursive

File: src/collections/utils.ts
Comment (amorton) [inline src/collections/utils.ts]: nit: remove unused code.

File: src/driver/collection.ts
Comment (amorton) [inline src/driver/collection.ts]: what should happen here, error or call something ?

File: src/driver/connection.ts
Comment (amorton) [inline src/driver/connection.ts]: we can prob remove all this, the path is the old doc V2 API path

**Review by amorton — CHANGES_REQUESTED** [review]
Looks good, mostly nit picking and tweaks. I went through with @kidrecursive  , he had a couple of points - we have both reviewed.

**Comment by kathirsvn** [comment]
> Heya, this looks good, is the plan to keep it in this repo until the document API is further along in stargate?

Hi @kidrecursive Yes, thats correct.

File: src/client/httpClient.ts
Comment (kathirsvn) [reply src/client/httpClient.ts]: Cleaned up the code.

File: src/client/httpClient.ts
Comment (kathirsvn) [reply src/client/httpClient.ts]: I set the message to be e.message. 

But I'm not clear on this part of the message. Could you please elaborate a little bit more on this?

> 'also add error:e to the object?'

File: src/client/httpClient.ts
Comment (kathirsvn) [reply src/client/httpClient.ts]: Thanks for the link. Changed console.log statements to logger statements as required.

File: src/collections/cursor.ts
Comment (kathirsvn) [reply src/collections/cursor.ts:29]: Changed it to filter

File: src/collections/cursor.ts
Comment (kathirsvn) [reply src/collections/cursor.ts]: Done

File: src/collections/cursor.ts
Comment (kathirsvn) [reply src/collections/cursor.ts]: Done

File: src/collections/cursor.ts
Comment (kathirsvn) [reply src/collections/cursor.ts]: Changed nextPageState to pageState and made it consistent

File: src/collections/db.ts
Comment (kathirsvn) [reply src/collections/db.ts]: Makes sense. Removed the catch block. It was there earlier to catch a specific HTTP error code when the server returns a when a create collection request is received and the collection is already present. But, it's no longer the scenario with the new server.

File: src/collections/utils.ts
Comment (kathirsvn) [reply src/collections/utils.ts]: Removed this function now. It's no longer needed I think. The only thing I was doing there now is to check if 'collation' is set in the options and throw an error if present. But I think, in our new implementation, server will return an error when 'collation' option is present in the request and not supported.

File: src/collections/utils.ts
Comment (kathirsvn) [reply src/collections/utils.ts]: Done

File: src/driver/collection.ts
Comment (kathirsvn) [reply src/driver/collection.ts]: Fixed this by adding an equivalent method in the internal collection.ts and invoking that from here. However, the findOneAndReplace in collection.ts is yet to be implemented since, it's not part of this iteration of the development

File: src/driver/connection.ts
Comment (kathirsvn) [reply src/driver/connection.ts]: Done

File: src/collections/cursor.ts
Comment (kathirsvn) [reply src/collections/cursor.ts]: Done

File: src/collections/collection.ts
Comment (vkarpov15) [inline src/collections/collection.ts]: In Mongoose, the expectation is that `insertOne()` will throw an error if the server returned an error. So this should probably `throw`.

File: src/driver/connection.ts
Comment (vkarpov15) [inline src/driver/connection.ts:98]: Is there any way for us to automatically create a keyspace before connecting? That would be very useful.

File: src/collections/collection.ts
Comment (kathirsvn) [reply src/collections/collection.ts]: Got it. A couple of questions.
1. Is that the expectation for all of the APIs? Can we throw a custom exception (something called `AstraException`) wrapping the error message and make it consistent across all of the APIs?
2. In which case, `acknowledged` is generally set to `false` vs throwing an error?

File: src/driver/connection.ts
Comment (kathirsvn) [reply src/driver/connection.ts:98]: Got it. There's no create keyspace operation allowed at this point. I added [an issue](https://github.com/stargate/sg-mongoose2/issues/11) for this to add this functionality when server side call is available.

File: src/collections/collection.ts
Comment (vkarpov15) [reply src/collections/collection.ts]: 1. In Mongoose 6, the expectation is that the driver layer will call the `callback()` with an error if the server threw an error. Mongoose does **not** expect the driver layer to return an object with an error description.
2. `acknowledged` is false if [write concern was disabled](https://www.mongodb.com/docs/manual/reference/method/db.collection.updateMany/#returns). In other words, if you asked the server for a "fire and forget" write where the server confirms it received the write operation, but doesn't confirm whether it persisted the write operation yet, then you'll get back `acknowledged: false`. `acknowledged: false` is not an error.

File: src/driver/connection.ts
Comment (vkarpov15) [reply src/driver/connection.ts:98]: That would be great. I don't think we need to create a keyspace for Astra, but for running stargate locally that would be very helpful.

Also, what do you think about generating an auth token automatically?

Comment (jeffreyscarpenter) [reply src/driver/connection.ts:98]: Would generating an auth token programmatically mean that we have to store credentials in the client like a username/password?

Comment (vkarpov15) [reply src/driver/connection.ts:98]: Yeah, in Mongoose code you typically pass database username/password to Mongoose, and Mongoose handles the auth process.

File: src/collections/collection.ts
Comment (kathirsvn) [reply src/collections/collection.ts]: @vkarpov15 
- I replaced all the `acknowledged: false` return statements (when there is a failure) with 'throw new Error' now. Please review and let me know if the latest change looks good with respect to error handling.
- And now I understand better on the `acknowledged` field in the response, thanks for the detailed explanation. This is going to be 'true' in our case here for all the APIs completed so far, since there is no fire and forget kind of APIs implemented so far.

File: src/collections/collection.ts
Comment (vkarpov15) [inline src/collections/collection.ts]: It would be great to consolidate this logic somewhere, like a `handleResponse()` function that checks `if (resp.errors && resp.errors.length > 0) { throw ... }`. We will likely want to change this error handling logic based on user feedback.

I would also recommend creating a separate `StargateServerError` class, something like:

```ts
class StargateServerError extends Error {
  constructor(response) {
    super(JSON.stringify(response.errors));
    this.errors = response.errors;
  }
}
```

That provides 1) a more informative error message, and 2) gives the user access to the unstringified errors.

File: src/client/httpClient.ts
Comment (vkarpov15) [inline src/client/httpClient.ts]: Aren't we calling this project stargate-mongoose?

Also, maybe pull the version number by importing `package.json`

Comment (jeffreyscarpenter) [reply src/client/httpClient.ts]: Both great ideas

Comment (kathirsvn) [reply src/client/httpClient.ts]: Fixed it to get both name and version from package.json

File: src/collections/collection.ts
Comment (kathirsvn) [reply src/collections/collection.ts]: @vkarpov15 Make sense. I created a common function to handle `errors` from server in the `httpClient.ts `itself, so we don't have to worry about error handling the individual APIs

Comment (kathirsvn) [reply src/collections/collection.ts]: Great suggestion on letting users access unstringified errors, thank you!

File: src/collections/collection.ts
Comment (vkarpov15) [inline src/collections/collection.ts]: Is there any way we can avoid this `as`? Otherwise, looks good so far :+1:

Comment (kathirsvn) [reply src/collections/collection.ts]: @vkarpov15 I changed it now by creating required object and returning instead of creating anonymous object and typecasting wherever possible. Please review and let me know if this looks good now.

**Comment by kathirsvn** [comment]
Hi @vkarpov15, does this PR look good now, are you planning for more review? Could you please approve if this looks good? I'll look for your comments if you are planning more review, we can wait, no issues. I just wanted to follow up on the approval if the review is complete.

**Review by vkarpov15 — APPROVED** [review]
Looks good so far. Will have to test some more as the backend gets more complete.

**Comment by kathirsvn** [comment]
Thanks @vkarpov15, I'll merge this PR once @amorton also approves this.
On the testing - Makes sense, please submit Issues in this repo, if you come across any issue during testing. I have already created some issues to work as next set of items.

**Review by amorton — APPROVED** [review]
discussed with Kathir, sounds like everything is good to go

**Pull request merged**
